arduino-ide-extension/src/browser/contributions/save-as-sketch.ts

```diff
@@ -84,91 +142,157 @@ export class SaveAsSketch extends SketchContribution {
     // If target exists, propose `directories.user`/${sketch.name}_copy_${yyyymmddHHMMss}
+    // IDE2 must never prompt an invalid sketch folder name (https://github.com/arduino/arduino-ide/pull/1833#issuecomment-1412569252)
     const defaultUri = containerDirUri.resolve(
-      exists
-        ? `${sketch.name}_copy_${dateFormat(new Date(), 'yyyymmddHHMMss')}`
-        : sketch.name
+      Sketch.toValidSketchFolderName(sketch.name, exists)
     );
     const defaultPath = await this.fileService.fsPath(defaultUri);
-    const { filePath, canceled } = await remote.dialog.showSaveDialog(
-      remote.getCurrentWindow(),
-      {
-        title: nls.localize(
-          'arduino/sketch/saveFolderAs',
-          'Save sketch folder as...'
-        ),
-        defaultPath,
+    return await this.promptLocalSketchFolderDestination(sketch, defaultPath);
+  }
+
+  /**
+   * Prompts for the new sketch folder name until a valid one is give,
+   * then resolves with the destination sketch folder URI string,
+   * or `undefined` if the operation was canceled.
+   */
+  private async promptLocalSketchFolderDestination(
+    sketch: Sketch,
+    defaultPath: string
+  ): Promise<string | undefined> {
+    let sketchFolderDestinationUri: string | undefined;
+    while (!sketchFolderDestinationUri) {
+      const { filePath } = await remote.dialog.showSaveDialog(
+        remote.getCurrentWindow(),
+        {
+          title: nls.localize(
+            'arduino/sketch/saveFolderAs',
+            'Save sketch folder as...'
+          ),
+          defaultPath,
+        }
+      );
+      if (!filePath) {
+        return undefined;
       }
-    );
-    if (!filePath || canceled) {
-      return false;
-    }
-    const destinationUri = await this.fileSystemExt.getUri(filePath);
-    if (!destinationUri) {
-      return false;
-    }
-    const workspaceUri = await this.sketchService.copy(sketch, {
-      destinationUri,
-    });
-    if (workspaceUri) {
-      await this.saveOntoCopiedSketch(sketch.mainFileUri, sketch.uri, workspaceUri);
-      if (markAsRecentlyOpened) {
-        this.sketchService.markAsRecentlyOpened(workspaceUri);
+      const destinationUri = await this.fileSystemExt.getUri(filePath);
+      // The new location of the sketch cannot be inside the location of current sketch.
+      // https://github.com/arduino/arduino-ide/issues/1882
+      let dialogContent: InvalidSketchFolderDialogContent | undefined;
+      if (new URI(sketch.uri).isEqualOrParent(new URI(destinationUri))) {
+        dialogContent = {
+          message: nls.localize(
+            'arduino/sketch/invalidSketchFolderLocationMessage',
+            "Invalid sketch folder location: '{0}'",
+            filePath
+          ),
+          details: nls.localize(
+            'arduino/sketch/invalidSketchFolderLocationDetails',
+            'You cannot save a sketch into a folder inside itself. This would go on forever.'
```

Comment:
```suggestion
            'You cannot save a sketch into a folder inside itself.'
```

The removed text is referencing a potential (it doesn't exist [anymore](https://github.com/arduino/arduino-ide/issues/381) in Arduino IDE 2.x) defect in the application which would cause it to create an infinitely recursive subfolder structure until something breaks in the file/operating system. So this text gives the impression that the restriction is only a workaround for a defect instead of intentional design. This action is prohibited because there isn't a valid use case for renaming to a location inside the source sketch and the user would only do this based on a misunderstanding of the function of the rename feature when applied to the primary sketch file.